extensions/ql-vscode/src/databases/local-databases-ui.ts

```diff
@@ -978,68 +977,86 @@ export class DatabaseUI extends DisposableObject {
    */
   private async chooseAndSetDatabase(
     byFolder: boolean,
-    subFolder: boolean,
     progress: ProgressCallback,
-  ): Promise<DatabaseItem[] | DatabaseItem | undefined> {
+  ): Promise<DatabaseItem | undefined> {
     const uri = await chooseDatabaseDir(byFolder);
     if (!uri) {
       return undefined;
     }
 
-    if (subFolder) {
-      if (!byFolder) {
-        return undefined;
-      }
+    if (byFolder && !uri.fsPath.endsWith("testproj")) {
+      const fixedUri = await this.fixDbUri(uri);
+      // we are selecting a database folder
+      return await this.databaseManager.openDatabase(fixedUri, {
+        type: "folder",
+      });
+    } else {
+      // we are selecting a database archive or a testproj.
+      // Unzip archives (if an archive) and copy into a workspace-controlled area
+      // before importing.
+      return await this.databaseFetcher.importLocalDatabase(
+        uri.toString(true),
+        progress,
+      );
+    }
+  }
 
-      const databases: DatabaseItem[] = [];
-      const failures: string[] = [];
-      const entries = await workspace.fs.readDirectory(uri);
-      for (const entry of entries) {
-        if (entry[1] === FileType.Directory) {
-          try {
-            const fixedUri = await this.fixDbUri(Uri.joinPath(uri, entry[0]));
-            const database = await this.databaseManager.openDatabase(fixedUri, {
-              type: "folder",
-            });
-            databases.push(database);
-          } catch (e) {
-            failures.push(entry[0]);
-          }
-        }
-      }
+  /**
+   * Ask the user for a parent directory that contains all databases.
+   * Returns all valid databases, or `undefined` if the operation was canceled.
+   */
+  private async chooseDatabasesParentFolder(
+    progress: ProgressCallback,
+  ): Promise<DatabaseItem[] | undefined> {
+    const uri = await chooseDatabaseDir(true);
+    if (!uri) {
+      return undefined;
+    }
 
-      if (failures.length) {
-        void showAndLogErrorMessage(
-          this.app.logger,
-          `Failed to import ${failures.length} database(s) (${failures.join(
-            ", ",
-          )}), successfully imported ${databases.length} database(s).`,
-        );
-      } else {
-        void showAndLogInformationMessage(
-          this.app.logger,
-          `Successfully imported ${databases.length} database(s).`,
-        );
+    const databases: DatabaseItem[] = [];
+    const failures: string[] = [];
+    const entries = await workspace.fs.readDirectory(uri);
+
+    for (const [index, entry] of entries.entries()) {
+      progress({
+        step: index + 1,
+        maxStep: entries.length,
+        message: `Importing ${entry[0]}`,
+      });
+
+      if (entry[1] === FileType.Directory) {
```

Comment:
So, this will ignore databases that are zips and it will also treat `testproj` databases as normal databases.

Probably the best approach would be to extract [lines 970 to 1001](https://github.com/github/vscode-codeql/pull/3797/files/1196902adc2b85b567cccc901600475d2cab956d..ce0b25feae469d7efd4f6aeef477f4dc6fb1213a#diff-24cce3c26831674245c5877c9683c8c8250c8595c4fa0d1eeece5c9b50bc5c8cR987-R1001) above into a separate function and call it hear and above.

Reply:
Good suggestion.